Add note on what methods are exported (#39362)
dotnet/docs: docs/core/deploying/native-aot/interop.md

The Native AOT compiler exports methods annotated with <xref:System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute> with a nonempty `EntryPoint` property as
public C entry points. This makes it possible to either dynamically or statically link the AOT compiled modules into external
programs. Only methods marked `UnmanagedCallersOnly` in the published assembly are considered: methods in project references or NuGet packages will not be exported.

> For more information, see [NativeLibrary sample](https://github.com/dotnet/samples/tree/main/core/nativeaot/NativeLibrary/README.md).

Reviewer: Should we also include information about using `UnmanagedEntryPointsAssembly` ItemGroup for exports from referenced assemblies?

Reviewer 2: @MichalStrehovsky ping

Author: We've not been documenting the internals of Xamarin bindings. I think there was an issue in the runtime repo on whether we want to do it and it was won't fixed.

Reviewer: I was referring to: https://github.com/dotnet/runtime/pull/83396 where we were introducing the `ItemGroup` for a more general use case (not just Xamarin support).

Author: Yep, and it was introduced for Xamarin. I don't know if this is ready for general consumption. For example we don't handle the situation when two UnmanagedCallersOnly methods have the same EntryPoint. This results in a bad crash. Allowing NuGets to add themselves to `UnmanagedEntryPointsAssembly` makes it more likely people are going to hit the bad failure modes.

We can easily detect conflicting entrypoints _within_ a single assembly with an analyzer (we don't have such analyzer, but it wouldn't be hard to make ont). We cannot detect a conflict when multiple assemblies are involved until it's too late. We don't have any recourse than to just fail the AOT compilation.

> programs. Only methods marked `UnmanagedCallersOnly` in the published assembly are considered: methods in project references or NuGet packages will not be exported.
> For more information, see [NativeLibrary sample](https://github.com/dotnet/samples/tree/main/core/nativeaot/NativeLibrary/README.md).

Reviewer 2: Just to note, we don't use hard break for sentences. We just let them wrap.

```suggestion
programs. Only methods marked `UnmanagedCallersOnly` in the published assembly are considered. Methods in project references or NuGet packages won't be exported.
For more information, see [NativeLibrary sample](https://github.com/dotnet/samples/tree/main/core/nativeaot/NativeLibrary/README.md).
```
